### ISSUE TITLE
Typing of toPromise is wrong

### ISSUE THREAD

**Issue by plol** [ReactiveX/rxjs#4196]
## Bug Report

**Current Behavior**
toPromise has the wrong type

**Reproduction**
```js
const p = of<string>().toPromise() // Type is Promise<string>
```

**Expected behavior**
p will in fact resolve to undefined and should have the type Promise<string | undefined>

**Comment by cartant** [comment]
This is a fairly subtle problem.

The promise will resolve when the observable completes - resolving with the last emitted value or with `undefined` if a value wasn't emitted.

However, the type declarations cannot express whether or not an observable will emita value, so it could be argued that the type for the promise should **always** be `Promise<T | undefined>`.

That would more correctly describe the behaviour, but it would be a breaking change.

**Comment by benlesh** [comment]
This was fixed in #5072
